File: day02/cube_conundrum.pl
#!/usr/bin/env perl
use v5.38;

sub good_set($cubes) {
    my %MAX_COUNT = (red => 12,
                     green => 13,
                     blue => 14,
                    );

    my @toks = split ', ', $cubes;
    for my $tok (@toks) {
        my ($count, $color) = split ' ', $tok;
        return 0 if $count > $MAX_COUNT{$color};
    }
    return 1;
}

sub good_game($game) {
    my @toks = split '; ', $game;
    for my $cubes (@toks) {
        return 0 unless good_set($cubes);
    }
    return 1;
}

sub min_set_prod($game) {
    my %min_set = map {$_ => 0} qw(red green blue);

    my @game_toks = split '; ', $game;
    for my $cubes (@game_toks) {
        my @cube_toks = split ', ', $cubes;
        for my $tok (@cube_toks) {
            my ($count, $color) = split ' ', $tok;
            $min_set{$color} = $count if $count > $min_set{$color};
        }
    }
    return $min_set{red} * $min_set{green} * $min_set{blue};
}

my $part1 = 0;
my $part2 = 0;
while (<>) {
    chomp;
    if (/Game (\d+): (.*)/) {
        my ($game_num, $game) = ($1, $2);
        $part1 += $game_num if good_game($game);
        $part2 += min_set_prod($game);
    }
}

say "Part 1: $part1";
say "Part 2: $part2";

File: day02/cube_conundrum.py
from sys import argv
import re

def good_set(cubes):
    MAX_COUNT = {
        'red': 12,
        'green': 13,
        'blue': 14,
    }

    toks = cubes.split(', ')
    for tok in toks:
        count, color = tok.split(' ')
        if MAX_COUNT[color] < int(count):
            return False

    return True

def good_game(game):
    toks = game.split('; ')
    for cubes in toks:
        if not good_set(cubes):
            return False

    return True

def min_set_prod(game):
    min_set = {k:0 for k in ('red', 'green', 'blue')}

    game_toks = game.split('; ')
    for cubes in game_toks:
        toks = cubes.split(', ')
        for tok in toks:
            count, color = tok.split(' ')
            min_set[color] = max(min_set[color], int(count))

    return min_set['red'] * min_set['green'] * min_set['blue']

part1 = 0
part2 = 0
with open(argv[1]) as f:
    for line in f:
        if m := re.match(r'Game (\d+): (.*)', line.rstrip()):
            game_num = int(m.group(1))
            game = m.group(2)
            if good_game(game):
                part1 += game_num
            part2 += min_set_prod(game)

print('Part 1:', part1)
print('Part 2:', part2)
